### PR TITLE
Write sbt.version to project/build.properties

### LINKED ISSUE #754: Print a warning if sbt.version is not defined in the project/build.properties
This would help creating more reproducible builds.

### PR DESCRIPTION
**Pull request by dwijnand**
Have sbt.version set in project/build.properties is a best practice
because it makes the build more deterministic and reproducible.

With this change sbt, after ensuring that the base directory is probably
an sbt project, writes out sbt.version in project/build.properties if it
is missing.

Fixes #754

### CODE REVIEW

File: main/src/main/scala/sbt/Main.scala [hunk 1]
```diff
@@ -13,25 +13,55 @@ import scala.annotation.tailrec
...
       )
     }
+
+  private val sbtVersionRegex = """sbt\.version\s*=.*""".r
+  private def isSbtVersionLine(s: String) = sbtVersionRegex.pattern matcher s matches ()
+
+  private def isSbtProject(baseDir: File, projectDir: File) =
+    projectDir.exists() || (baseDir * "*.sbt").get.nonEmpty
+
+  private def setSbtVersion(baseDir: File, sbtVersion: String) = {
```
Comment (eed3si9n) [inline main/src/main/scala/sbt/Main.scala:43]: We might want to put opt-out Java flag around this method.

**Review by eed3si9n — CHANGES_REQUESTED** [review]

**Review by eed3si9n — APPROVED** [review]

File: main/src/main/scala/sbt/Main.scala [hunk 2]
```diff
@@ -13,25 +13,56 @@ import scala.annotation.tailrec
...
+    val buildPropsLines = if (buildProps.canRead) IO.readLines(buildProps) else Nil
+
+    val sbtVersionAbsent = buildPropsLines forall (!isSbtVersionLine(_))
+
+    if (sbtVersionAbsent) {
+      val errorMessage = s"WARN: No sbt.version set in project/build.properties, base directory: $baseDir"
+      try {
+        if (isSbtProject(baseDir, projectDir)) {
+          val newBuildPropsLines = s"sbt.version=$sbtVersion" :: buildPropsLines
+          IO.writeLines(buildProps, newBuildPropsLines)
```
Comment (Duhemm) [inline main/src/main/scala/sbt/Main.scala]: I think sbt should tell the user that it's writing the file. Software should not perform side effects silently without explicit user consent.

Comment (eed3si9n) [reply main/src/main/scala/sbt/Main.scala]: Between not booting up at all and creating `build.properties`, creating `build.properties` I think is lesser of two bad scenarios.

Comment (Duhemm) [reply main/src/main/scala/sbt/Main.scala]: I agree. My point is that sbt should also say "don't worry, I'll write it out for you".

Comment (eed3si9n) [reply main/src/main/scala/sbt/Main.scala]: ok. I don't have a strong opinion on displaying warning or silently creating the file. We can go either way.

Comment (dwijnand) [reply main/src/main/scala/sbt/Main.scala]: Oh good, because I agree with Martin.

**Pull request merged**